khtools/extract_coding.py

```diff
@@ -0,0 +1,510 @@
+    "low_complexity_peptide":
+    "peptide sequence from low "
+    "complexity (low entropy) translated"
+    " reads"
+}
+SCORING_DF_COLUMNS = [
+    'read_id', 'jaccard_in_peptide_db', 'n_kmers', 'classification'
+]
+
+
+def write_fasta(file_handle, description, sequence):
+    file_handle.write(f">{description}\n{sequence}\n")
+
+
+def open_and_announce(filename, seqtype, quiet=False):
+    if not quiet:
+        announcement = SEQTYPE_TO_ANNOUNCEMENT[seqtype]
+        click.echo(f"Writing {announcement} to {filename}", err=True)
+    return open(filename, 'w')
+
+
+def three_frame_translation(seq, debug=False):
+    if debug:
+        warning_filter = 'default'
+    else:
+        warning_filter = 'ignore'
+
+    with warnings.catch_warnings():
+        warnings.simplefilter(warning_filter)
+        for frame in range(3):
+            translation = seq[frame:].translate()
+            yield translation
+
+
+def three_frame_translation_no_stops(seq, debug=False, sign=1):
+    """Remove translations with stop codons & keep track of reading frame"""
+    return {
+        sign * (i + 1): t
+        for i, t in enumerate(three_frame_translation(seq, debug))
+        if '*' not in t
+    }
+
+
+def six_frame_translation_no_stops(seq, debug=False):
+    forward_translations = three_frame_translation_no_stops(seq, debug)
+
+    # Sign=-1 sets the reading frames as negative to make it obvious they are
+    # from the reverse strand
+    reverse_translations = three_frame_translation_no_stops(
+        seq.reverse_complement(), debug, sign=-1)
+    forward_translations.update(reverse_translations)
+    return forward_translations
+
+
+def score_single_translation(translation,
+                             peptide_bloom_filter,
+                             peptide_ksize,
+                             molecule='protein',
+                             verbose=True):
+    encoded = encode_peptide(translation, molecule)
+    kmers = list(set(kmerize(str(encoded), peptide_ksize)))
+    hashes = [hash_murmur(kmer) for kmer in kmers]
+    n_kmers = len(kmers)
+    n_kmers_in_peptide_db = sum(1 for h in hashes
+                                if peptide_bloom_filter.get(h) > 0)
+    if verbose > 1:
+        click.echo(f"\ttranslation: \t{encoded}", err=True)
+        click.echo("\tkmers:", ' '.join(kmers), err=True)
+
+    if verbose > 1:
+        kmers_in_peptide_db = {(k, h): peptide_bloom_filter.get(h)
+                               for k, h in zip(kmers, hashes)}
+        # Print keys (kmers) only
+        click.echo(f"\tK-mers in peptide database:", err=True)
+        click.echo(kmers_in_peptide_db, err=True)
+
+    fraction_in_peptide_db = n_kmers_in_peptide_db / n_kmers
+
+    return fraction_in_peptide_db, n_kmers
+
+
+def evaluate_is_fastp_low_complexity(seq, complexity_threshold=0.3):
+    """Use fastp's definition of complexity
+
+    fastp prpject: https://github.com/OpenGene/fastp
+
+    Low complexity = Many runs of the same base in a row
```

Comment:
can you give an example here is AAAAA low complexity ATAAAA has some complexity?